interpreter/src/Language/Haskell/Interpreter.hs

```diff
@@ -43,41 +46,70 @@ data CompilationError
     deriving stock (Show, Eq, Generic)
     deriving anyclass (ToJSON, FromJSON)
 
+data InterpreterError
+    = CompilationErrors [CompilationError]
+    | TimeoutError Text
+    deriving stock (Show, Eq, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+newtype SourceCode = SourceCode Text
+   deriving stock (Generic)
+   deriving newtype (ToJSON, FromJSON)
+   deriving anyclass (Newtype)
+
 -- | spawn an external process to runghc a file
 --
 --   If you set the environmental varaiable GHC_BIN_DIR
 --   then the executable runghc in that path will be used.
 --   This is useful if you want to your file to be run with some packages
 --   available, you can create a wrapper runghc that includes these
 --
---   Any errors are converted to [CompilationError]
+--   Any errors are converted to InterpreterError
 runghc
-    :: (MonadIO m, MonadError [CompilationError] m, MonadMask m)
-    => [String]
+    :: (Show t, TimeUnit t, MonadIO m, MonadError InterpreterError m, MonadMask m)
+    => t
+    -> [String]
     -> FilePath
     -> m String
-runghc runghcOpts file = do
+runghc t runghcOpts file = do
```

Comment:
Call me picky, but can we rename `t` to `timeout` or `timeoutValue` or something?